[Merged by Bors] - feat(Topology): some `IsOpenMap` lemmas

peabrainiac opened the pull request:
Some lemmas for `IsOpenMap` that I needed. I've also added variants for `IsClosedMap` to maintain symmetry in the API.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

peabrainiac committed:
upstream `IsOpenMap` lemmas

peabrainiac committed:
add variants for `IsClosedMap`

peabrainiac commented:
This also exposes some inconsistency in the naming: `Continuous.restrict` and `IsOpenMap.restrict` already both exist, but the former is about `Set.MapsTo.restrict` while the latter is about `Set.restrict`. I've opted to name the `Set.MapsTo.restrict` lemma for `IsOpenMap` added here `IsOpenMap.restrict_mapsTo` because `ContinuousOn` similarly uses `.restrict` and `.restrict_mapsTo`, but it might be good to standardise that more across different function properties in the future.

peabrainiac committed:
merge master

j-loreaux commented on Mathlib/Topology/Constructions.lean:
```diff
@@ -365,18 +365,48 @@ theorem IsClosed.isClosedMap_subtype_val {s : Set X} (hs : IsClosed s) :
+    IsClosedMap (s.restrict f) :=
+  hf.comp hs.isClosedMap_subtype_val
+
 @[continuity, fun_prop]
 theorem Continuous.subtype_mk {f : Y → X} (h : Continuous f) (hp : ∀ x, p (f x)) :
     Continuous fun x => (⟨f x, hp x⟩ : Subtype p) :=
   continuous_induced_rng.2 h
 
+theorem IsOpenMap.subtype_mk {f : Y → X} (hf : IsOpenMap f) {s : Set X} (hfs : ∀ x, f x ∈ s) :
+    IsOpenMap fun x ↦ (⟨f x, hfs x⟩ : Subtype s) := fun u hu ↦ by
+  convert (hf u hu).preimage continuous_subtype_val
+  exact Set.ext fun _ ↦ exists_congr fun _ ↦ and_congr_right' Subtype.ext_iff
+
+theorem IsClosedMap.subtype_mk {f : Y → X} (hf : IsClosedMap f) {s : Set X} (hfs : ∀ x, f x ∈ s) :
```
Can you please add `Subtype.coind` versions of these?

j-loreaux commented on Mathlib/Topology/Constructions.lean:
```diff
@@ -365,18 +365,48 @@ theorem IsClosed.isClosedMap_subtype_val {s : Set X} (hs : IsClosed s) :
+    IsClosedMap fun x ↦ (⟨f x, hfs x⟩ : Subtype s) := fun u hu ↦ by
+  convert (hf u hu).preimage continuous_subtype_val
+  exact Set.ext fun _ ↦ exists_congr fun _ ↦ and_congr_right' Subtype.ext_iff
+
 theorem Continuous.subtype_map {f : X → Y} (h : Continuous f) {q : Y → Prop}
```
```suggestion

@[fun_prop]
theorem Continuous.subtype_map {f : X → Y} (h : Continuous f) {q : Y → Prop}
```

j-loreaux commented on Mathlib/Topology/Constructions.lean:
```diff
@@ -365,18 +365,48 @@ theorem IsClosed.isClosedMap_subtype_val {s : Set X} (hs : IsClosed s) :
+theorem IsClosedMap.subtype_map {f : X → Y} (hf : IsClosedMap f) {s : Set X} {t : Set Y}
+    (hs : IsClosed s) (hst : ∀ x, s x → t (f x)) : IsClosedMap (Subtype.map f hst) :=
+  (hf.comp hs.isClosedMap_subtype_val).subtype_mk _
+
 theorem continuous_inclusion {s t : Set X} (h : s ⊆ t) : Continuous (inclusion h) :=
```
```suggestion

@[fun_prop]
theorem continuous_inclusion {s t : Set X} (h : s ⊆ t) : Continuous (inclusion h) :=
```

peabrainiac replied:
I don't think this is necessary: `Set.inclusion` is an abbrev, so `fun_prop` can see through it and use other lemmas to prove continuity - in fact, the proof of this lemma could already be replaced with `by fun_prop`. Let me know if you think it makes sense to tag it with `fun_prop` anyways though.

j-loreaux reviewed (commented):
> I've opted to name the `Set.MapsTo.restrict` lemma for `IsOpenMap` added here `IsOpenMap.restrict_mapsTo` because `ContinuousOn` similarly uses `.restrict` and `.restrict_mapsTo`, but it might be good to standardise that more across different function properties in the future.

The correct name according to our conventions is actually `mapsToRestrict`. I think the only lemmas with the incorrect name are [LipschitzOnWith.to_restrict_mapsTo](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/EMetricSpace/Lipschitz.html#LipschitzOnWith.to_restrict_mapsTo) and [ContinuousOn.restrict_mapsTo](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/ContinuousOn.html#ContinuousOn.restrict_mapsTo)

Would you like to instead standardize these to `mapsToRestrict` and add the appropriate deprecations?

peabrainiac committed:
merge master

peabrainiac committed:
rename `restrict_mapsTo`-lemmas to `mapsToRestrict`

peabrainiac committed:
apply suggestions

peabrainiac commented:
> The correct name according to our conventions is actually `mapsToRestrict`. I think the only lemmas with the incorrect name are [LipschitzOnWith.to_restrict_mapsTo](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/EMetricSpace/Lipschitz.html#LipschitzOnWith.to_restrict_mapsTo) and [ContinuousOn.restrict_mapsTo](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/ContinuousOn.html#ContinuousOn.restrict_mapsTo)
> 
> Would you like to instead standardize these to `mapsToRestrict` and add the appropriate deprecations?

I've renamed these lemmas now, and also added the variants for `Subtype.coind` you requested (I'm not aware of any lemmas like this for `Subtype.coind` in the library already, but adding them does seem reasonable - I've named them `subtype_coind` because of the similar variants `subtype_mk` and `subtype_map`). I'm curious though - it seems to me that if lemmas about `MapsTo.restrict` should be called `mapsToRestrict`, these lemmas should be called e.g. `subtypeMk` too instead of `subtype_mk` Are their names just outdated, or is there some other reason that they carry the name they have?

j-loreaux commented:
Thanks!

bors merge
